dashboard/app/models/contact_rollups_processed.rb

```diff
@@ -301,6 +302,32 @@ def self.extract_form_roles(contact_data)
     return uniq_form_roles.blank? ? {} : {form_roles: uniq_form_roles}
   end
 
+  def self.extract_state(contact_data)
+    # The priority is: school state > user state > form geo state
+    # US state in schools table is in abbreviation, must convert it back to state name.
+    school_state = extract_field_latest_value contact_data, 'dashboard.schools', 'state'
+    if school_state
+      state_name = get_us_state_from_abbr(school_state, true) || school_state
+      return {state: state_name}
+    end
+
+    user_state = extract_field_latest_value contact_data, 'dashboard.users', 'state'
```

Comment:
This confused me for a while, because our User model doesn't have a column or method named `state`.  I finally noticed that [ContactRollupsRaw includes user_geos](https://github.com/code-dot-org/code-dot-org/blob/7a99855caef4dcf160c178e7b00405980952af13/dashboard/app/models/contact_rollups_raw.rb#L121-L141) when it populates the `dashboard.users` source.

Reply:
We think of `user_geos` table as an extension of `users` table. So instead of doing 2 separate extraction, reading user_id from `users` table, and reading user location from `user_geos` (still has to join with `users` to get email address), we just do one join and get both user_id and user location.

Reply:
:+1: Sounds good!